# Проект по SQL.

## Цели и задачи исследования:

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных. В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

## Описание данных:

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

## Задания:
- Посчитайте, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитайте количество обзоров и среднюю оценку;
- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

## План работы:

- Цели и задачи исследования;
- Исследование таблиц — вывод первых строк;
- Сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитать количество обзоров и среднюю оценку;
- Определить издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
- Определить автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
- Посчитать среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine
from IPython.display import Image

In [2]:
# ER-Диаграмма
Image(url='https://concrete-web-bad.notion.site/image/https%3A%2F%2Fs3-us-west-2.amazonaws.com%2Fsecure.notion-static.com%2F069818d1-0e5c-4d87-a461-0de584ab9c33%2FUntitled_(33).png?id=bd53c8db-b4fd-49eb-8cc9-572ebb3c9163&table=block&spaceId=9e4bd47b-c6e6-4ca3-bcee-279794b47315&width=2000&userId=&cache=v2')

In [3]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

In [4]:
# Функция выполнения sql запроса

def func(sql):
    return pd.read_sql(sql,con=engine)

### Вывод первых строк таблиц

In [5]:
# для книг
sql_books = '''
SELECT *
FROM books
'''
books = func(sql_books)
books.head()

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [6]:
# для авторов
sql_authors = '''
SELECT *
FROM authors
'''
authors = func(sql_authors)
authors.head()

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [7]:
# для рейтингов
sql_ratings = '''
SELECT *
FROM ratings
'''
ratings = func(sql_ratings)
ratings.head()

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [8]:
# для обзоров
sql_reviews = '''
SELECT *
FROM reviews
'''
reviews = func(sql_reviews)
reviews.head()

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [9]:
# для издательств
sql_publishers = '''
SELECT *
FROM publishers
'''
publishers = func(sql_publishers)
publishers.head()

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

### Сколько книг вышло после 1 января 2000 года;

In [10]:
query = '''
SELECT COUNT(book_id)
FROM books 
WHERE publication_date > '2000-01-01'
'''

func(query)

count
0    819

***После 1 января 2000 года вышло 819 книг***

### Для каждой книги посчитать количество обзоров и среднюю оценку;

In [11]:
query = '''
SELECT books.book_id as id,
       books.title as title,
       AVG(ratings.rating) as avg_rating,
       COUNT(DISTINCT reviews.review_id) as reviews_count
FROM books
LEFT JOIN reviews on reviews.book_id = books.book_id
LEFT JOIN ratings on ratings.book_id = books.book_id
GROUP BY title, id
ORDER BY reviews_count DESC
'''
func(query)

id                                              title  avg_rating  \
0    948                            Twilight (Twilight  #1)    3.662500   
1    963                                Water for Elephants    3.977273   
2    734                                   The Glass Castle    4.206897   
3    302  Harry Potter and the Prisoner of Azkaban (Harr...    4.414634   
4    695  The Curious Incident of the Dog in the Night-Time    4.081081   
..   ...                                                ...         ...   
995   83    Anne Rice's The Vampire Lestat: A Graphic Novel    3.666667   
996  808                            The Natural Way to Draw    3.000000   
997  672   The Cat in the Hat and Other Dr. Seuss Favorites    5.000000   
998  221                          Essential Tales and Poems    4.000000   
999  191  Disney's Beauty and the Beast (A Little Golden...    4.000000   

     reviews_count  
0                7  
1                6  
2                6  
3                6  
4                6  
..             ...  
995              0  
996              0  
997              0  
998              0  
999              0  

[1000 rows x 4 columns]

***Вывели список книг по количеству обзоров с указанием рейтинга. Книга с id 948 имеет наибольшее количество обзоров***

### Определить издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;

In [12]:
query = '''
SELECT pub.publisher as publisher_name,
       COUNT(DISTINCT books.book_id) as books_count
FROM publishers as pub
 LEFT JOIN books on pub.publisher_id = books.publisher_id
WHERE books.num_pages > 50
GROUP BY publisher_name
ORDER BY books_count DESC
LIMIT 10;
'''
func(query)

publisher_name  books_count
0             Penguin Books           42
1                   Vintage           31
2  Grand Central Publishing           25
3          Penguin Classics           24
4          Ballantine Books           19
5                    Bantam           19
6                   Berkley           17
7             Berkley Books           14
8        St. Martin's Press           14
9                     Delta           13

***Издательство Penguin Books выпустило 42 книги толще 50 страниц. Вывели топ-10 издательств.***

### Определить автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;

In [13]:
query = '''
SELECT authors.author,
       ROUND(AVG(ratings.rating),2) AS avg_rating
                   
FROM ratings            
LEFT JOIN books AS books ON ratings.book_id=books.book_id
LEFT JOIN authors AS authors ON books.author_id=authors.author_id
WHERE ratings.book_id IN (SELECT ratings.book_id
                          FROM ratings
                          GROUP BY ratings.book_id
                          HAVING COUNT(ratings.book_id) >= 50)
GROUP BY authors.author
ORDER BY avg_rating DESC
LIMIT 1
'''
func(query)

author  avg_rating
0  J.K. Rowling/Mary GrandPré        4.29

***Автор с самой высокой средней оценкой книг - J.K. Rowling/Mary GrandPré со средним рейтингом 4.288.***

### Посчитать среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [14]:
query = '''
WITH t as
    (SELECT username,
            COUNT(review_id) as review_count
     FROM reviews
     WHERE username in (SELECT username
                         FROM ratings
                         GROUP BY username
                         HAVING COUNT(rating) > 50)
    GROUP BY username)
    
SELECT AVG(review_count)
FROM t
       
'''
func(query)

avg
0  24.333333

***Среднее количество обзоров от пользователей, которые поставили больше 50 оценок равно 24.33***

## Общий вывод

- После 1 января 2000 года было выпущено 819 книг.
- По количеству обзоров на первом месте книга Twilight со средним рейтингом 3.66.
- Издательство, выпустившее наибольшее количество книг, - издательство "Penguin Books".
- Автор с самой высокой средней оценкой книг - J.K. Rowling/Mary GrandPré.
- Среднее количество обзоров от пользователей, поставивших более 50 оценок, - 24.33